Completed all tasks [team member 3] Game class

binodbhusal opened the pull request:
**The following tasks were accomplished in this milestone**

- [x] Create a Game class in a game.rb file.

- [x] Create an Author class with an association to the Item class 

- [x] Implemented methods:
         add_item method in the Author class
         take an instance of the Item class as an input
         add the input item to the collection of items
        add self as a property of the item object (by using the correct setter from the item object)
        can_be_archived?() in the Game class
        override the method from the parent class

- [x] Add unit tests for all implemented methods.

- [x] The following options  available:

           List of games
           List all authors (e.g. 'Stephen King')
           Add a game

- [x] All data  preserved by saving collections in .json files.

- [x] Create a schema.sql file with tables that will be analogical to the structure 

             games table (add all properties and associations from the parent Item class as table columns)
             authors table

binodbhusal committed:
Add author class

binodbhusal committed:
Add game class

binodbhusal committed:
Add GameApp class

binodbhusal committed:
Add unit test for all Author class methods

binodbhusal committed:
Add unit test for all Game class methods

binodbhusal committed:
Add games.json

binodbhusal committed:
Add authors.json

binodbhusal committed:
add details

binodbhusal committed:
add sql query to create author and game tables

binodbhusal committed:
Modify files

binodbhusal committed:
Fix test case Rspec issue

luigirazum commented on game/author.rb:
```diff
@@ -0,0 +1,27 @@
+    @last_name = last_name
+    @items = []
+    super({})
+  end
```
Hey @binodbhusal, remember that Author class inherits from Item class.
So it should receive the full params for calling the Item constructor with super and the params for initialize the Author instance as well.
Check out the constructor at the Item class
```rb
def initialize(params)
    @id = params[:id] || rand(1..10_000)
    @genre = params[:genre]
    @author = params[:author]
    @source = params[:source]
    @label = params[:label]
    @publish_date = create_date(params[:publish_date])
    @archived = params[:archived] || false
  end
```
We also need to use the OR assignment for default values like `params[:archived] || false`, so if you send the `archived` params it is assigned that value, if not, the `false` default value is set.

luigirazum commented on game/author.rb:
```diff
@@ -0,0 +1,27 @@
+    return if @items.include?(item)
+
+    @items << item
+    item.author = self
```
There is a method called `add_author` in the Item class, we should use that method to set the association between Author and Item.

luigirazum commented on game/game.rb:
```diff
@@ -0,0 +1,31 @@
+    @can_be_archived = false
+    @publish_date = Date.today
+    @authors = []
+  end
```
- Same case for this constructor, we can change it by the params style, like in the Item class.
- can_be_archived and publish_date are inherited properties from Item.
- authors shouldn't exist here, there is no authors in the UML diagram.

luigirazum commented on game/game.rb:
```diff
@@ -0,0 +1,31 @@
+  def associate_with_author(author)
+    author.add_item(self)
+    @authors = [author] # Store the association with the author
+  end
```
I think there is no need to create this `associate_with_author` method... we can use author.add_item.
authors doesn't exist in game, only an author that it is inherited from Item

luigirazum commented on game/game.rb:
```diff
@@ -0,0 +1,31 @@
+    @authors = [author] # Store the association with the author
+  end
+
+  def can_be_archived?
+    super && (Date.today - @played_at_date).to_i / 365 > 2
+  end
```
`can_be_archived?` it's overriden in Game class, it is defined originally at Item class, and it is a private method, so it must be a private method in Game class too.

luigirazum commented on game/gameapp.rb:
- There was already an `app.rb` file that has the operations defined, just insert your code in the appropriate methods and add the instance variables you require like authors and games.

luigirazum commented on game/gameapp.rb:
```diff
@@ -0,0 +1,112 @@
+    save_games
+  end
+
+  def list_all_games
+    puts 'List of games:'
+    @games.each_with_index do |game, index|
+      puts "#{index + 1}. Multiplayer: #{game.multiplayer}, Last Played Date: #{game.played_at_date}"
+    end
+  end
+
+  def list_all_authors
+    puts 'List of Authors:'
+    @authors.each do |author|
+      puts "ID: #{author.id}. FullName: #{author.first_name} #{author.last_name}"
+    end
+  end
+
```
**👀 OPTIONAL**
Perhaps it would be good to have a validation at the beginning of the list_all methods to check if we have authors or games to list we print them, if not we show a message saying there is no games/authors yet.

luigirazum commented on game/author.rb:
I think it would be good to move the classes in `game` folder into the `classes` folder that we discuss initially.

luigirazum reviewed (changes requested):
Hi @binodbhusal <br>Good job so far! <br>👏👏👏 
---
Status: ♻️ Changes Required
===
![the-simpsons-homer-simpson](https://user-images.githubusercontent.com/28307960/157369993-50d5ba12-099f-4f2a-9a19-5e57ee2d1c96.gif)

There are some issues that you still need to work on to go to the next project, but you're almost there!
## To Highlight 🎉
- No Linters errors ✔️
- The app works as expected 👏 ✔️

 ## Required Changes ♻️
- [ ] The changes I think it would be good to apply are in the commented code.
- [ ] **_Please 👀 Check the comments below_**

## Optional suggestions ✍️
- There is an optional suggestion, it would be good if you take it into account.
_Every comment with the <kbd>OPTIONAL</kbd> prefix is not crucial enough to stop the approval of this PR. <br>However, I strongly recommend you take them into account as they can make your code better._

***
Cheers and Happy Coding! 👨‍💻
---
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me `@luigirazum` in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. <br>You should use the same Pull Request submitted for the first review,<br>either valid or invalid unless it is requested otherwise._** 

----
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

binodbhusal committed:
include add_author method item class asspciation

binodbhusal committed:
include add-author method from item for association

binodbhusal committed:
include inheritance and update as per review

binodbhusal committed:
update file

binodbhusal committed:
update files and test cases

luigirazum reviewed (approved):
STATUS: 🟢 APPROVED!!!  🎉 🎉 🎉
===
Hi @binodbhusal, 🙋‍♂️ I'm Luis your code reviewer. 
---
Good work on making the changes requested in the previous review.
### Great job !!! 👏 👏 👏 
Your project is **complete!**

Now you can merge this branch and move on 🥇 
---
Congratulations! 🎉 🎉 🎉 :champagne: :champagne: :champagne:
===
## To Highlight 🎉
- No Linters errors ✔️
- You can create a Game ✔️
- You can create an Author ✔️
- The app works as expected ✅
- Good commit messages. 👍

![giphy](https://github.com/otmaneechchafyky/ruby-capstone/assets/67355479/713228bf-3837-4afe-8f69-c0baa9a0eb8a)

## Optional suggestions ✍️
- N/A
> _Every comment with the <kbd>OPTIONAL</kbd> prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

***
Happy Coding! 👨‍💻 and Good Luck 🍀 with the coming ones 🤞
---
Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me `@luigirazum` in your question so I can receive the notification, you can also reach me out on Slack using the same tag as well.

> **_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._** 

----
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

luigirazum committed:
Merge app from game-class branch

luigirazum committed:
Fix 'Author' class, it doesn't has a Parent class

luigirazum committed:
Fix 'Game' class issues.
- Parent Item class is in the same folder.
   - require_relative '../item'
- can_be_archived is a method
   - @can_be_archived = params[:can_be_archived] || false
- publish_date is defined in the Parent class
   - @publish_date = create_date(params[:publish_date])
- There are no @authors in the UML diagram for this class.
   - @authors = []
- The association is done when you call author.add_item(item)
   - def associate_with_author(author)

luigirazum committed:
Adapt 'GameClass' to work with the project

luigirazum committed:
Fix calls to create game with author

luigirazum committed:
Fix tests for Author and Game classes

luigirazum committed:
Add json files test data
